### PR TITLE
Richards benchmark for CHERI

### PR DESCRIPTION
**Pull request by jeremysinger**
Contributing Richards benchmark ported to CHERI. This code runs on Morello/CheriBSD.

### CODE REVIEW

File: richards-benchmark/README.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+Richards is an OS task scheduler simulation benchmark, designed for 
+comparing system implementation languages. The benchmark was originally
+implemented in BCPL by Martin Richards.
+See https://www.cl.cam.ac.uk/~mr10/Bench.html for details.
+This is a CHERI-compatible version of the Richards benchmark, with
+optional support for BoehmGC. The code is derived from Stefan Marr's
+classic benchmarks repository at
+https://github.com/smarr/Classic-Benchmarks
+
+The Richards benchmark does not have a licence. Instead, Martin Richards
```
Comment (ltratt) [inline richards-benchmark/README.md:12]: I checked this with Martin in 2012 and he said:

> Yes, everything in my bench distribution is in the public domain except for the directory sunjava.

so it does have a licence (which is good!).

File: richards-benchmark/harness.h [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+ * It is included by every benchmark file and provides the time measuring,
+ * as well as standardized command-line argument parsing.
+ */
+
+#include <sys/time.h>
+
+unsigned long microseconds() {
+  // Not monotonic
+  struct timeval t;
+  gettimeofday(&t, NULL);
```
Comment (ltratt) [inline richards-benchmark/harness.h:12]: Should we use `clock_gettime`? Admittedly, `CLOCK_MONOTONIC` on Linux isn't actually monotonic in the commonly understood sense and we need to use `CLOCK_MONOTONIC_RAW` on Linux (but not BSD).

**Comment by ltratt** [comment]
I wasn't quite sure if/what we'd had to change to CHERIfy this?

**Comment by jeremysinger** [comment]
> I wasn't quite sure if/what we'd had to change to CHERIfy this?

I made minimal changes ... see this diff:

8a9,13
> **  Modified by:  Jeremy Singer, 25 May 2022
> **    adapted to use uintptr_t for compatibility with architectures
> **    that do not have word-sized pointers (e.g. CHERI) 
> **    also adapted to use Boehm-Demers-Weiser GC instead of system
> **    malloc, when compiled with -DGC
20a26
> #include <stdint.h>
24a31,37
> #ifdef GC
>   #include "gc.h"
>   #define MALLOC GC_MALLOC
> #else
>   #define MALLOC malloc
> #endif // GC
> 
82,83c95,96
<     long            t_v1;
<     long            t_v2;
---
>     uintptr_t       t_v1;
>     uintptr_t       t_v2;
91,92c104,105
< long    v1;
< long    v2;
---
> uintptr_t v1;
> uintptr_t v2;
105,106c118,119
<                 long v1,
<                 long v2)
---
>                 uintptr_t v1,
>                 uintptr_t v2)
108c121
<     struct task *t = (struct task *)malloc(sizeof(struct task));
---
>     struct task *t = (struct task *)MALLOC(sizeof(struct task));
125c138
<     struct packet *p = (struct packet *)malloc(sizeof(struct packet));
---
>     struct packet *p = (struct packet *)MALLOC(sizeof(struct packet));
305c318
<       v1 = (long)(((struct packet *)v1)->p_link);
---
>       v1 = (uintptr_t)(((struct packet *)v1)->p_link);
313c326
<       v2 = (long)(((struct packet *)v2)->p_link);
---
>       v2 = (uintptr_t)(((struct packet *)v2)->p_link);
333c346
<         v1 = (long)pkt;
---
>         v1 = (uintptr_t)pkt;

File: richards-benchmark/README.md (hunk 1 above)
Comment (jeremysinger) [reply richards-benchmark/README.md:12]: good ... so do we need a LICENSE.txt ? What should it say?

**Comment by ltratt** [comment]
If you're feeling brave, you can inject a git commit in the with the unmodified benchmark, and then a second commit with your diff in, so that it's easier to see what's changed. I won't blame you if that sounds a bit fiddly though :)

File: richards-benchmark/harness.h (hunk 2 above)
Comment (jeremysinger) [reply richards-benchmark/harness.h:12]: I didn't change this code, I retained Stefan Marr's original timing harness from https://github.com/smarr/Classic-Benchmarks/

File: richards-benchmark/README.md (hunk 1 above)
Comment (ltratt) [reply richards-benchmark/README.md:12]: Good idea! We can just say "Placed into the public domain by Martin Richards."

**Comment by jeremysinger** [comment]
> If you're feeling brave ...

I shall rise to your challenge...

File: richards-benchmark/harness.h (hunk 2 above)
Comment (ltratt) [reply richards-benchmark/harness.h:12]: Oh! We should reference that (as it's copyrighted by Stefan, or his employers).

**Comment by jeremysinger** [comment]
> > If you're feeling brave ...
> 
> I shall rise to your challenge...

have I done the injection you suggested, @ltratt ?

**Comment by ltratt** [comment]
Works for me! Shall we add a LICENSE commit and delete the "The Richards benchmark does not have a licence." paragraph? With that, I think we're ready to merge!

**Comment by ltratt** [comment]
I would suggest:

> The Richards benchmark has been placed into the public domain by Martin Richards.
>
> The benchmark runner is ...

Actually we need to ask @smarr what the license of his runner is. Stefan... help :)

**Comment by smarr** [comment]
Ehm, you're talking about this here? https://github.com/smarr/Classic-Benchmarks/commits/master/benchmarks/harness.h (URLs in Git commit messages can be very helpful when they are there, especially when attributing from where and when stuff was copied).

I'd hope that this harness.h doesn't reach the level of requiring copyright.

Otherwise, your choice:

 - Option 1: Stefan needs to be coauthor on all papers (for any future lawyers: this is clearly an attempt at a joke)
 - Option 2: what ever is convenient to you, if that's Public Domain to match Richard's that's fine

**Comment by ltratt** [comment]
Yep, that's copyrightable! Public domain works for me: @jeremysinger shall we go with the wording "The Richards benchmark has been placed into the public domain by Martin Richards. The benchmark harness has been placed into the public domain by Stefan Marr".

> Option 1: Stefan needs to be coauthor on all papers (for any future lawyers: this is clearly an attempt at a joke)

:p

**Comment by jeremysinger** [comment]
done @ltratt

**Comment by ltratt** [comment]
Ah, you either need to turn clang-format off for these files (probably the right thing to do here) or format them according to its whims.

**Comment by jeremysinger** [comment]
> Ah, you either need to turn clang-format off for these files 
do I add a .clang-format file to this directory, or use `// clang-format off` comments in the files?

**Comment by ltratt** [comment]
You know as much (or, probably more accurately for the both of us, as little :)) as I do. @drpdr might have a suggestion?

**Comment by drpdr** [comment]
@jeremysinger I'd suggest adding a `.clang-format` file and running `clang-format` (check the version used in the case you have further errors).

**Comment by drpdr** [comment]
> @jeremysinger I'd suggest adding a `.clang-format` file and running `clang-format` (check the version used in the case you have further errors).

Oh no, wait. This is in `cheri-examples`, you just have to run `clang-format`.

**Comment by drpdr** [comment]
Oh, sorry. My bad. So `// clang-format off` should do the trick.

**Comment by jeremysinger** [comment]
bors try

**Pull request merged**